### PR TITLE
surface_profiling

### PR DESCRIPTION
**Pull request by tianqisong0117**
surface_profiling

### CODE REVIEW

File: slam/surface_profiling.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
+
+import numpy as np
+import slam.geodesics
+import trimesh
+
+
+def cortical_surface_profiling(mesh, theta, r_step, m, save_path):
+    """
+
```
Comment (gauzias) [inline slam/surface_profiling.py:9]: please provide description

File: slam/surface_profiling.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        else:
+            # the origin point is out of intersection points
+            points_sign = list(np.zeros([m, 3]))
+
+        profile_list.append(points_sign)
+
+    return profile_list
+
+
+def vert2poly_indices(vertex_array, poly_array):
```
Comment (gauzias) [inline slam/surface_profiling.py:151]: there are some tools in trimesh for such purpose, plaese look at trimesh/utils

Comment (gauzias) [reply slam/surface_profiling.py:151]: this kind of fuctions should be moved to the slam/utils.py to ease re-use

**Review by gauzias — COMMENTED** [review]
So after discussion, I will merge the PR from Julien and refactor a bit so that you can the update (pull upstream) to get the hinge surface generation functions that will help you for the example script

File: slam/surface_profiling.py (hunk 1 above)
Comment (JulienLefevreMars) [reply slam/surface_profiling.py:9]: Important also to have correct names for the variables. For instance, mesh, r_step, save_path are relatively straightforward but theta or m far less.

File: slam/surface_profiling.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        # get the profile samplings on surface
+        sam_prof = surface_profiling_vert(vert_i, vert_norm_i, theta, r_step, m, intersect_mesh)
+
+        # compute the 2D coordinates (x, y) for all profile points
+        sample_x, sample_y = compute_profile_coord_x_y(np.array(sam_prof), vert[i], norm[i])
+
+        profile_samples_x.append(sample_x)
+        profile_samples_y.append(sample_y)
+
+        print(i, 'is done')
```
Comment (JulienLefevreMars) [inline slam/surface_profiling.py:54]: print can be done in a VERBOSE mode (False by default)

File: slam/surface_profiling.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    :param v_n: normal vector of v
+    :type: ndarray
+    :param v_p: vector projected
+    :type: ndarray
+    :return: v_t
+    """
+    unitev_n = v_n / np.linalg.norm(v_n)
+
+    coeff_v_pn = np.dot(v_p, v_n) / np.linalg.norm(v_n)
+
```
Comment (JulienLefevreMars) [inline slam/surface_profiling.py:213]: redundant:
use unitev_n in line 212 instead of v_n

File: slam/surface_profiling.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    coeff = coeff_v_pn.reshape([coeff_v_pn.size, 1])
+
+    v_pn = coeff * unitev_n
+
+    return v_pn
+
```
Comment (JulienLefevreMars) [inline slam/surface_profiling.py:242]: Those three last functions could be added in a basic_geometry/utils module ?

**Review by JulienLefevreMars — COMMENTED** [review]
First attempt to do a review. Ok for lines 0-80

File: slam/surface_profiling.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+import numpy as np
+import slam.geodesics
```
Comment (alexpron) [inline slam/surface_profiling.py:3]: PEP compliance about imports, slam import should be after numpy import separated by one line (pep8 ll deal with i)

File: slam/surface_profiling.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+def get_rotate_matrix(rot_axis, angle):
+    """
+    for a pair of rotation axis and angle, calculate the rotate matrix
+    :param rot_axis: rotation axis
+    :param angle: rotation angle
+    :return: rotate matrix of [3, 3]
+    """
+
+    # normalize the rotate axis
+    r_n = rot_axis / np.linalg.norm(rot_axis)
```
Comment (alexpron) [inline slam/surface_profiling.py]: Possible runtime error if rot_axis is null vector (norm equal 0)

**Comment by gauzias** [comment]
@all-contributors please add @tianqisong0117 for code Examples Ideas

**Pull request merged**